### PR TITLE
Enhance Template Number example

### PR DESCRIPTION
**Pull request by ferbar**
## Description:

Just enhancing the example a bit.

## Checklist:

  - [x] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [ ] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

**Comment by oxan** [comment]
I would prefer it if the default example keeps using optimistic mode, as that's (in my experience) what most people actually want.

**Comment by ferbar** [comment]
@oxan: If I'm right you are either using the template number with optimistic + initial_value + set_action, when you won't read a value or you are using it with lambda + set_action when you want feedback. If so, I would suggest 2 examples since they give a better clue what can be done with the template number.

```
# Example configuration entry for a write - only number:
number:
  - platform: template
    name: "Template number"
    optimistic: true
    initial_value: 50
    min_value: 0
    max_value: 100
    step: 1
    set_action:
      [something]

# Example configuration entry, currend value read by a lambda function
  - platform: template
    name: "Changeable Global Variable"
    min_value: 0
    max_value: 100
    step: 1
    lambda: |-
      [something]
    set_action:
      [something]
```

**Comment by oxan** [comment]
Yeah, two examples seems like a good idea.

**Comment by ferbar** [comment]
@oxan I found out that the best usage for the Template Number is probably the add_on_state_callback function in the C++ code which I noted in the example. Have a look at it please. Thanks!

File: components/number/template.rst [hunk 1]
```diff
@@ -10,14 +10,32 @@ using :ref:`lambdas <config-lambda>`.
 
 .. code-block:: yaml
 
-    # Example configuration entry
+    # Example configuration entry for setting a value, either in combination with the set_action configuration or the add_on_state_callback() method in a custom component.
```
Comment (jesserockz) [inline components/number/template.rst:13]: I don't like this addition exposing information for custom components here.
Anyone developing custom components should be looking at the api docs themselves, not the configuration docs.

**Pull request closed without merging**